### PR TITLE
Instance: Allow a stateful VM to be started even if its root disk `size.state` parameter is not set

### PR DESCRIPTION
**Pull request by gabrielmougard**
This allows a user to do:

```
lxc launch ubuntu:jammy v1 --vm --config migration.stateful=true -s local
```

Whereas it was only working with a `remote` storage backend before.
The root disk's `size.state` related error will only arise when a user attempts to do a stateful operation like `lxc stop v1 --stateful`

### CODE REVIEW

File: lxd/instance/drivers/driver_qemu.go [hunk 1]
```diff
@@ -1040,8 +1040,9 @@ func (d *qemu) saveState(monitor *qmp.Monitor) error {
 	return nil
 }
 
-// validateRootDiskStatefulStop validates the state of the root disk before stopping the instance.
-func (d *qemu) validateRootDiskStatefulStop() error {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1044]: Why did you rename this, rather than only calling it during a stateful stop?

Now its being called during start up and not doing anything?

**Review by tomponline — CHANGES_REQUESTED** [review]
Not quite there yet im afraid.

File: lxd/instance/drivers/driver_qemu.go (hunk 1 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:1044]: It still need to check if the root disk device exists, that's why I kept it

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1044]: Lets just do that separately rather then bundle them together.

File: lxd/instance/drivers/driver_qemu.go [hunk 2]
```diff
@@ -1054,28 +1054,31 @@ func (d *qemu) validateRootDiskStatefulStop() error {
 		return err
 	}
 
-	stateDiskSizeStr := pool.Driver().Info().DefaultVMBlockFilesystemSize
-	if rootDiskDevice["size.state"] != "" {
-		stateDiskSizeStr = rootDiskDevice["size.state"]
-	}
+	// Only local root disks need additional checks.
+	if !pool.Driver().Info().Remote {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1058]: This isnt correct. If you're statefully stopping a VM, it doesn't matter if its running on a remote pool or not, it still needs somewhere to save its state file. So these checks should always be done.

**Comment by tomponline** [comment]
@gabrielmougard please can you describe the testing process you used for this, before I merge it? Thanks

**Comment by tomponline** [comment]
> The root disk's `size.state` related error will only arise when a user attempts to do a stateful operation like `lxc stop v1 --stateful`

And this allows them to set size.state whilst the VM is running too?

**Comment by gabrielmougard** [comment]
> Thanks.
> 
> Please can you also add tests for this to https://github.com/canonical/lxd-ci/blob/main/tests/storage-vm to start a VM, then try to stop it statefully without setting size.state, confirm its rejected, and then try and set size.state whilst its running and confirm it will statefully stop and then statefully start again.
> 
> Thanks

@tomponline Here it is https://github.com/canonical/lxd-ci/pull/74

**Comment by tomponline** [comment]
@gabrielmougard please can you respond to comments in your PRs saying "fixed" and then resolve the thread so i can see what is resolved and what is outstanding, thanks.

File: lxd/instance/drivers/driver_qemu.go (hunk 2 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go]: fixed

File: lxd/instance/drivers/driver_qemu.go [hunk 3]
```diff
@@ -1097,17 +1100,10 @@ func (d *qemu) validateStartup(stateful bool, statusCode api.StatusCode) error {
 	// Otherwise, there will not be enough disk space to write the instance state to disk during any subsequent stops.
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go]: fixed

**Comment by tomponline** [comment]
thanks!
